### PR TITLE
🐛 Fix validation warning for unions of Literal and other type

### LINKED ISSUE #6601: Serialization warning with `tp.Literal[False] | str`
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

Seems like a slightly different serialization check bug to #6467
```
UserWarning: Pydantic serializer warnings:
  Expected `Union[literal[False], str]` but got `bool` - serialized value may not be as expected
  return self.__pydantic_serializer__.to_json(
```

### Example Code

```Python
import typing as tp
from pydantic import BaseModel

class Bar(BaseModel):
    bar: tp.Literal[False] | str

Bar(bar=False).model_dump_json()
```

### Python, Pydantic & OS Version

```Text
pydantic version: 2.0.2
        pydantic-core version: 2.1.2 release build profile
                 install path: venv/lib/python3.11/site-packages/pydantic
               python version: 3.11.4 (main, Jun  7 2023, 12:45:48) [GCC 11.3.0]
                     platform: Linux-5.19.0-46-generic-x86_64-with-glibc2.35
     optional deps. installed: ['devtools', 'email-validator', 'typing-extensions']
```

Selected Assignee: @lig

### PR DESCRIPTION
**Pull request by lig**
## Change Summary

Reproduce the issue

## Related issue number

Fix #6601 

## Checklist

* [x] The pull request title is a good summary of the changes - it will be used in the changelog
* [x] Unit tests for the changes exist
* [x] Tests pass on CI and coverage remains at 100%
* [x] Documentation reflects the changes where applicable
* [x] My PR is ready to review, **please add a comment including the phrase "please review" to assign reviewers**

Selected Reviewer: @hramezani

### CODE REVIEW

**Comment by lig** [comment]
depends on https://github.com/pydantic/pydantic-core/pull/783

**Comment by lig** [comment]
please review

File: tests/test_edge_cases.py [hunk 1]
```diff
@@ -2563,3 +2563,29 @@ class MyModel(TypedDict):
...
+)
+def test_union_literal_with_other_type(literal_type, other_type, data, json_value, data_reversed, json_value_reversed):
+    class Model(BaseModel):
+        value: Union[literal_type, other_type]
+        value_types_reversed: Union[other_type, literal_type]
+
+    from pprint import pprint
+
+    pprint(Model.__pydantic_core_schema__)
+    pprint(Model.model_json_schema())
```
[the remark this answers is not in the document]
Comment (lig) [reply tests/test_edge_cases.py:2587]: Thanks!

**Comment by lig** [comment]
please review

**Review by hramezani — APPROVED** [review]

**Pull request merged**